{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af9f4d82-f712-4316-97f1-b04572a3eb4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "import gffutils\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "194602b2-6975-4939-85f2-3b1ca5acde5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME_PATH = \"../datasets/schizosaccharomyces_pombe/GCF_000002945.2_ASM294v3_genomic.fna\"\n",
    "GFF_PATH = \"../datasets/schizosaccharomyces_pombe/genomic.gff\"\n",
    "ORGANISM = \"fission_yeast\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e103b3a-e9c4-4d08-933e-05dd3cf0a951",
   "metadata": {},
   "source": [
    "### Load Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e5f1cb5-86af-42d5-8659-031663da6413",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading genome...\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "print('loading genome...')\n",
    "#reading entire genome into memory - this would NOT work if we use a big genome like humans\n",
    "genome = SeqIO.to_dict(SeqIO.parse(GENOME_PATH, \"fasta\"))\n",
    "print('done!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f153a43-e0fb-4221-8bac-b4d5d3f32126",
   "metadata": {},
   "source": [
    "#### Genome exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0577cb24-df25-42b5-b6e9-a861c14c149c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_003424.3': SeqRecord(seq=Seq('GATCACGTACATCACCTTGTAAGAATTTATCTGCAATAGTCCTTCGGTATTGTA...ATC'), id='NC_003424.3', name='NC_003424.3', description='NC_003424.3 Schizosaccharomyces pombe strain 972h- genome assembly, chromosome: I', dbxrefs=[]),\n",
       " 'NC_003423.3': SeqRecord(seq=Seq('GATCTCGCAACTCTTAATAAAGCTAATTCCTGCTAATTCGCTATACACTAAATC...TTT'), id='NC_003423.3', name='NC_003423.3', description='NC_003423.3 Schizosaccharomyces pombe strain 972h- genome assembly, chromosome: II', dbxrefs=[]),\n",
       " 'NC_003421.2': SeqRecord(seq=Seq('GATCAGCCAAAATGGCTGATCCAGCTATTTAGCAGGTTAAGGTCTCGTTCGTTA...GGG'), id='NC_003421.2', name='NC_003421.2', description='NC_003421.2 Schizosaccharomyces pombe strain 972h- genome assembly, chromosome: III', dbxrefs=[]),\n",
       " 'NC_088682.1': SeqRecord(seq=Seq('AATGTGTAATTTGAATCAATTTCTATCCTTTTAAAAATACATCATAAAGTGTAC...TGT'), id='NC_088682.1', name='NC_088682.1', description='NC_088682.1 Schizosaccharomyces pombe isolate MT1 mitochondrion, complete genome', dbxrefs=[])}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "586d3539-7cfb-4743-90af-1099c9a18594",
   "metadata": {},
   "source": [
    "### Build GFF DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7e914f56-99fa-4418-82a8-905876108d8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading existing GFF database...\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(f\"{ORGANISM}_genome.db\"):\n",
    "    print(\"creating GFF database...\")\n",
    "    db = gffutils.create_db(\n",
    "        GFF_PATH,\n",
    "        dbfn=f\"{ORGANISM}_genome.db\",\n",
    "        force=True,\n",
    "        keep_order=True,\n",
    "        merge_strategy=\"merge\",\n",
    "        sort_attribute_values=True\n",
    "    )\n",
    "else:\n",
    "    print(\"loading existing GFF database...\")\n",
    "    db = gffutils.FeatureDB(f\"{ORGANISM}_genome.db\", keep_order=True)\n",
    "print(\"done!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb662773-1fd5-4fb6-8bab-f852e65efdfc",
   "metadata": {},
   "source": [
    "#### DB Exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "17b2898d-58ee-469a-a218-a43b8953e1c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CDS', 'LONG_TERMINAL_REPEAT', 'RNase_P_RNA', 'exon', 'five_prime_UTR', 'gene', 'intron', 'lnc_RNA', 'mRNA', 'ncRNA', 'pseudogene', 'rRNA', 'region', 'sequence_feature', 'snRNA', 'snoRNA', 'tRNA', 'three_prime_UTR']\n"
     ]
    }
   ],
   "source": [
    "print([f for f in db.featuretypes()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9c546ef1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_003424.3\tRefSeq\tCDS\t1\t5662\t.\t-\t0\tID=cds-NP_595040.1;Dbxref=GeneID:2541932,GenBank:NP_595040.1,PomBase:SPAC212.11;Name=NP_595040.1;gbkey=CDS;gene=tlh1;locus_tag=SPOM_SPAC212.11;partial=true;start_range=.,1;Parent=rna-NM_001018168.1;product=RecQ type DNA helicase;protein_id=NP_595040.1\n",
      "GATCACGTACATCACCTTGTAAGAATTTATCTGCAATAGTCCTTCGGTATTGTACATTGTTCCAAGCATAGTAAACTAACGATATCAAGTTTGCCTTTCTAGCCCATGACCTACAGTCAGAAGTGTAAGCCATATCACTGTCGGCATGTTCAAACTTTGTCAAACcacaaaataaacaaagtcCTTGAAATCGAATACGTAGTTTACATTCTCGCAAGTTGTGGTCGGCCTTGCCACATTTATAACAAGTAGATAAGCGTACGGGGCATGCTTTCCCGGTATGAGCACGAATTTCTGTGTCTGGGTTACCAAGAGTGCAACTTAGACATTCATCTTTATACACTCGAAAGAATCGCTCGAGTTGTGCTGAAATGTCAGTTGAGTCTACCCATTGTTTTTTGACGGCTTGCACACGACTTTTATACTCGGCGAAATGAATGGTACGAGAATCAACATCACCATCACCATCACCATCACCAATACCAATACCAATACCAACAATACCACTCATACCTCCAACACTACCATGAACACGATTCATGTCATGGTTGACGCCCTGTTGTACATTGTTCAAATGTTCATCGTATGTGCGTTTATGACCACGACGATGGCTGAGCGACATGTCTTCCAAACCATACTTTAGATTCGTATTAGCATTGCCAAAGCTTTTCTTCCATGTTGGCGATACAGACGAAGGGAACAAAGTGATTTCCGATAATTCCGACAAAGACACTGGCGTAGTACTAGTAGTGTTAGTGTTCATAGCACTCATACCACTGCTATTCCCTGGCTGTGGTGGGGGGGAAGACGAAAACGATGCATTATAACGCGAATGTGTTGCAATGGCTGGTTTCGGTGTTTCTGGCAATGTCGGTTTCACTCCATACATCGTAGACACAGTTGATTCTTCACCAAGTAACGAATCTGAGCATCTTGAGCAGTAAACACAGTTAGCAAACGATGCACAACATACACATTCGCCATCCATTTCACTTGCGAGAAACGATCGAACACACATCgtattatcattaagaaagtttttcatCGAATCCTCTACGTAGCTCGACCATGTAGAATCATATTTCTCGTAAAACAATGCTGCAATCGCATACTTGCCATCTCTTCCAGCTCGACCTGTCTCCTGTACATAATCCATAGATGAAGCTGGTAATCCATAGTGTACTACTAAACGCACTCCCATATAGTTGATACCGAGTCCGAATGCCTTGGTAGCGATCATGATTCGTGTCTTCCCATTTGCATTTCGAAACGCGTCAAAGTTCATTTGTCGTTCTTCGTCGCTTACATCTCCTGTGTAAATGGTTACATGGGTGTGAGCGAACAAGTCCGACTGGTGAAGACGACGATGAATGTATTCGACATCCTTTTTGGTTCGACAAAAGATGATAGCACGTCCATCACCTTCAAAGACCTTTGTTCGTTTCATCAACGTTCGCAAGTCCAACAAAAACTCGGTATTATCATAGGGATgcaaaaagtaaaagatgTTTTCCCGTGCAGTAGACGTTTCTCGTAAAACGTAAAAGTTTGTACAAAACGTTTGTCTGGCAACCATCTCTAGTTGCCTGGGAAAGGTGGCACTCAACAAGTGCAATGGTGCATACAAGCCAGACAACCGCGATGCTCTCGACAAAGCCGTTCTCCATGCTCCACTGGTGAGTAACAAATGTGCTTCATCAATTACCACTCGTGCCAAACGACCAAGTGTTGCcaaactttcaaaaaaccGAAGACCACTGTTGGTTAATGCTGACTCGTATGTCAAGATAAACAAATCGGGAAGTTGCGTCTCTAAAGTCAATCGTACATCTTTGAATGCGGTCCAATTCCCCGAACAAACAAGCAGTCCCTTTTCATTCACTCTGAGCATCATATCTTGCCGTAACGACATCATGGGCACTAGAACCAGCGTGACCATATTCATCACCTTTCCTGGggtttgtctttttttttcgatgAGCGCCGGTATCAAAAACGACAAAGACTTTCCACCTCCAGTTGGGAGTACGGTAATCAAATTCATACGatttaaaagagaaaaataaaccGATTGAAATTGTTTCAGCGAACGGAATTTCGCTTCGAGACCATAATATTGAGACAATGCCCAACAACAATGTTGAACAGAAGTTTGAGATGGTTGAAGAAGTCCTGGGCGGTAGTGAAATGCGGTGGGCGCGGTGGGCGTAGTGGCCAATTGGTCTTGTAAGTCATGAGAGTCTTGATTTACAACGGGCTGTGTCATTTCAGCCTTTACCTCGTCCTTGTTCTTCTCCTCATTTGTCTTGTCCTTTTCCTTGTCCCCCTCCACCATTTGTTCACCTACCAACCTTGCCAGCTTCTCATTCAGCGCTTCCTCCAATTGATCAACTCGCTTGATGAATGGGTGCTTTGTTTCCACCAACAGTCCATGGGTCGGGTTATCTCGAATCTGTAATAGTTCCTGCCACTTATAGCTTGCACGAAAAAAGTTGGCGAAATAATCCGACGGCAGATAATGCAAGTTGTCCATAGTGCGCCCATAGATGTAGAGCGCTGTGTTTCGTGTATGTCCAGCCTGTAAATCGAAATATGATTCCCTCGTCATTTCTTCCTCgatgtttttttctttaaaatagtGAGCAATGTGCCTGTAGTTTCGAAATCCCAACGGTTTTTGAATGCATTGGTAGGTGGcctttggaaaaattcGATACGGTAAATCTCTTTGCAATCGTTCGCCAGCAATCACAAACATGAAATCCAAGTATACAGCTACTTTCGACCTATCAGCGGTTGTCACATACTTCATCAATGCTTCCAATGGTCGAACCAAAACATAGTACCGAAGTGCTAAAATGGAAAGCGGCTCAGAAAGAAACCTGGGGATTGGCTTTTCAGCAAACTTCATATTACGAGTCTTATCGTATCTGCTGTAAATCATCAGCCTTCCAAACATCAAATACAATTCGCGAGTCTTATACTTGCCATTCCGCAAAGTCCAATACACCATCTCTTGTGCTCTGGCTGGTTGTCCAGCAGATATGTACACTAGACAAAACAATAACTCATTGAAGATAGATGCCTTTGTCtcatactttttaatagcaGATTTAGACCAATGTAATTTGAGAAACACATCATTGTCATTATCATTAGAGTTGTCGTTGTAGTGGTTGCTATAGTTGTCGCACTCATTTCTTTGCTGATCACTTACTCCACCAAACAACTTTGTCAGATCCTTCTTAGACATCGATGGGAAAAACCGATGACAAAGTTCACTGTTCTTAAGTATTGTTCGTAACAGCACACTATGAGATGAACGTATCTCTATCGACTCtttaaagaatgaatatCCCAATTTACTATTGTTTATGTCATCTCCAATTGACCTTAGTAACGAAATGGGTAATAGTTTTTCCACATTCATATCAAAACATAGCTCTTTTAACAAGTCCTGAACTTTGGCAAACATCCCCTCATACAGAGCTGTCAGCTCACTCAAGTCCAATCGCACCATGTTATTGAATGCATCTCCAATGCACGGGTACATAGAATTACTTCGTGGCTGAAACAATACGGTACTTGGGCTTAGTCCTTGCAGTGTATTTGCATACCCTTCCGTTCGAATGATATCCCTAACGACTTTTTGGGTTTCGTAAAGAACATCatataaatttctttgaccAAGCGTGATAAAGCAAAACACATGAGCATAAGCAATAGAACCAGCGGTTTGAAACGATGGAATGGACAACAAGtcgatttgtttttcatcgATCAATCGTTGTAACTCACATGCTGCTACAAGACGACACGTGTATATTAGGGGTGCAAAGCAGGCAGAGATTTCATAGGCATATTTTGAAGTACCATCTTTATGCAAAGACATGCATGCAAGAAACTCCATGATTTCCAAGTCTTGaacatttttagaaaaatcgTATTGTTGCAGAAAAGCCAATTTTAATGCCTCGTGCAGTTCCAAGTATGCCAGCTTATCATCATCGTCAGTATTGGTGTTGTTATTGTTATTGTTATTGTTACTGTCTCTGTCATTGTCATTGTCATTGTCATTGTCGTCTTCAGCACTATCAAAGTTGGCGACGTTGCTGATGTTACTGCTGCTGTTACTGCCACCTTCTTTTCTCTCCACTGTTGCTCCACACATTGCTTCCAAGTGTCTTCTTATCCCCGAGTCCGCTGATGGACGACGTAACAGAAATATTAGGTATAGTGCTGCACTTTGAGCATATCTCACAACTGTATCCTTTTGCTCTAAGCGACAAAGTCCTCGCTCTTGTGCACTATAGCTTCCTCCTTGAGTAAACATTCGTCTCATGCCCATGGTCATTCCACTTAGATGACGGAATCCTCGATAGTAATAACGAATCAACTGGTTTACAATGGTATCCAAAGTTACAATTCCATTCGTATATTCGAATCGAACCAGCTGATCCCAATTTTCTCCAAGTTTTTTCACAATTGCCGTCCAATTGAATTGttgatttaaaagatttttctCTCGTGCATTGACAAtctcttcctcttcctcttcatcgttatctttttcatcatcttcaAGAATAGTGTTAGTGTTGTTGTCTGTTGCTTGGTCAACATCATCTTGtcgttttttattttcaccTTTGTCTTGCTGCTCCCGCCGCTGTTGATTCTTCTccataaattttgaaaagcaaaagtcCACCATTCTTTCCTTGATGTTTTGTGCATCTTCTTTCTCCTCCCCCtcctcctcttcttcttcttctcctCCATTCAAAGGTACATAATCGACTTTGAAGAACTGACATCGTTGAGCATACTTGTTCTTAACCGTTTGCAAAGCTGTTCGACGTATAcaattttccaattttatAACTTTTCCATGAGTACGTCGAACATGATGGCGAAAGGTATCCATGATTGCATGTACGAATCCTGTTCCGTTTTTTGTGCACGGCACACACTCATATCCATTTAAAAGTACTGGTAGTCCTCTGATGTATGGGTATACATGGGTTTGAGAAGAATGAGTTTGTAAAACATTTGTTGGACTTTTTAACTTTAAAGTCCTTAGTTCTTGAAACCACAAAAGATCTTCGTTTAACTCTAGTTTATGCACAATTTGCATATGTTGCGCAGTGTGAATCACGTTTAACAAGCATTCACATTCTACGCACATCAAAGCATGAATAGACAAAATGGAGAGTCCATAATTAGCCAACTTGTGATTCAAGTCAGTATAATTTGATGATGGGCTGTTGTCCAATTTGGTGTCATTTGTACTTTGCCAAATTAGCCGTGCAGTCCACAAGTCGTCATTTGCACTCGTAGTGAGAGTAGAGAGAGCAGAGAGAGCAGAGAGAGCGGGTAGTTGACGCTCCTTGGAAGAATTGCAAGCCTCCTCACTGGTAGCCTCACTGGCAGCCTCAACGATGATTGGCTTATCCTTTTTCTCGTCCTGAACGAAAGAGCTGGGTTCATAGACGACGATATCATCGTCTTTACAAGTTTCGTCTCTGTATCGACATTCCTTGAAATGTTGAACGATCCTTTCTACATTATCAAATTGAGTTCCACATTGACAAGTAAAGCATCCGGCGATATCTCTAGCAGTTTTTGAAGCGACTTTAGCAATTTCTGAAGCGACGACCAT\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "feat = next(db.features_of_type('CDS'))\n",
    "print(feat)\n",
    "feature_sequence = genome[feat.seqid].seq[feat.start - 1 : feat.end]\n",
    "print(feature_sequence)\n",
    "print(f'{len(feature_sequence) == (feat.end - feat.start + 1)}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "882ab4a3-e0f8-4747-8965-f1e72f2a38cb",
   "metadata": {},
   "source": [
    "### Parse data from DB into .CSV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "985813fb",
   "metadata": {},
   "source": [
    "#### Get genome sequence and state sequence\n",
    "For the purpose of this project, we define a coding sequence ONLY as a sequence labeled with `CDS` in the GFF file. Everything else we consider to be non-coding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fb2b06b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get cds intervals\n",
    "cds_intervals = defaultdict(list)\n",
    "\n",
    "for cds in db.features_of_type('CDS'):\n",
    "    cds_intervals[cds.seqid].append((cds.start, cds.end))\n",
    "\n",
    "len(cds_intervals.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2b044d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get coding and noncoding sequences\n",
    "sequences = []\n",
    "state_mask = []\n",
    "\n",
    "for seqid, seqinfo in genome.items():\n",
    "    chromosome = seqinfo.seq\n",
    "\n",
    "    labels = [False] * len(chromosome)\n",
    "\n",
    "    for start, end in cds_intervals[seqid]:\n",
    "        for i in range(start - 1, end):\n",
    "            labels[i] = True\n",
    "    \n",
    "    base_seq = \"\".join([b for b in chromosome])\n",
    "    label_seq = \"\".join(['1' if c else '0' for c in labels])\n",
    "\n",
    "    sequences.append(base_seq)\n",
    "    state_mask.append(label_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ac97a0a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "#entire genome in my memory\n",
    "entire_genome_sequence = \"\".join(sequences).upper()\n",
    "entire_state_sequence = \"\".join(state_mask).upper()\n",
    "print(len(entire_genome_sequence) == len(entire_state_sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0834b10",
   "metadata": {},
   "source": [
    "#### Parse into k-mers and CSV\n",
    "Will be using 3-mers but this can easily be changed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e43c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = 3\n",
    "with open(f\"{ORGANISM}_{k}-mers.csv\", \"w\") as f:\n",
    "    f.write(\"3mer,coding\\n\")\n",
    "    for idx in range(len(entire_genome_sequence) - k + 1):\n",
    "        #assuming state depends on middle nucleotide (this assumption is kinda whacky)\n",
    "        #maybe later want to just discard k-mers at transitions between states but for now this is fine?\n",
    "        f.write(f\"{entire_genome_sequence[idx : idx + k]},{entire_state_sequence[idx + k//2]}\\n\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
